velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -728,6 +728,94 @@ struct ArrayRemoveNullFunctionString {
   }
 };
 
+template <typename T>
+struct ArrayNGramsFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+
+  // Fast path for primitives.
```

Comment:
Do we need this after https://github.com/facebookincubator/velox/pull/8194 ?

CC: @laithsakka

Reply:
we still need it because of this part of the code which does not use add_items .
mainly to avoid calling copy_from on newGranularItem when newGranularItem is pritmive.

this can be simplified potentially by support start and end index in add items or ranges
ex:
out.addItems(input, 0 , n); 
instead of the loop bellow it can be useful also for functions like array_trim. 

```
or (auto j = 0; j < n; ++j) {
        if (input[i + j].has_value()) {
          auto& newGranularItem = newItem.add_item();
         // Ma
          newGranularItem = input[i + j].value();
        } else {
          newItem.add_null();
        }
      }
  ```

Reply:
@laithsakka Thank you for clarifying.

> mainly to avoid calling copy_from on newGranularItem when newGranularItem is pritmive.

Copying primitive values is cheap, no? Why is it important to avoid calling copy_from for these?

Reply:
In this particular function, I wonder if bulk of the cost goes into resizing the array, since we do not provide an estimate on the final size, no?